### PR TITLE
bootloader: add helper for creating a bootloader based on gadget

### PR DESCRIPTION
**Pull request by bboozzoo**
The bootloader.Find() expects bootloader configuration to be present in the root
directory. However, when we are installing a run mode system, the bootloader
config is not present until after InstallBootConfig() gets called, which happens
only after the filesystem has been created and populated. However, when using
the ContentObserver to track trusted assets, we must know what the bootloader
is.

Add a helper for creating a bootloader based on what is carried by the gadget.
Move the basic checks for config file existence to Find().

### CODE REVIEW

File: bootloader/bootloader.go [hunk 1]
```diff
@@ -327,3 +322,15 @@ func removeKernelAssetsFromBootDir(bootDir string, s snap.PlaceInfo) error {
 
 	return nil
 }
+
+// ForGadget returns a bootloader matching
```
Comment (pedronis) [inline bootloader/bootloader.go:326]: the doc comment is not complete

File: bootloader/bootloader.go [hunk 2]
```diff
@@ -327,3 +322,15 @@ func removeKernelAssetsFromBootDir(bootDir string, s snap.PlaceInfo) error {
...
+func ForGadget(gadgetDir, rootDir string, opts *Options) (Bootloader, error) {
```
Comment (pedronis) [inline bootloader/bootloader.go:327]: in a follow up we can use this in InstallBootConfig

File: bootloader/export_test.go [hunk 3]
```diff
@@ -43,7 +43,7 @@ func MockAndroidBootFile(c *C, rootdir string, mode os.FileMode) {
 	c.Assert(err, IsNil)
 }
 
-func NewUboot(rootdir string, blOpts *Options) ExtractedRecoveryKernelImageBootloader {
+func NewUboot(rootdir string, blOpts *Options) Bootloader {
```
Comment (pedronis) [inline bootloader/export_test.go:46]: we could keep the signature the do the cast here mabye?

File: bootloader/export_test.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func MockUbootFiles(c *C, rootdir string, blOpts *Options) {
 	c.Assert(err, IsNil)
 }
 
-func NewGrub(rootdir string, opts *Options) RecoveryAwareBootloader {
+func NewGrub(rootdir string, opts *Options) Bootloader {
 	return newGrub(rootdir, opts)
```
Comment (pedronis) [inline bootloader/export_test.go:65]: same

File: bootloader/bootloader.go [hunk 5]
```diff
@@ -234,9 +234,18 @@ func InstallBootConfig(gadgetDir, rootDir string, opts *Options) error {
 	return fmt.Errorf("cannot find boot config in %q", gadgetDir)
 }
 
+type bootloaderNewFunc func(rootdir string, opts *Options) Bootloader
+
 var (
 	forcedBootloader Bootloader
 	forcedError      error
+
+	bootloaders = []bootloaderNewFunc{
```
Comment (pedronis) [inline bootloader/bootloader.go:243]: I would  give this its own var close to the type definition of the function with a comment like

// bootloaders list all possible bootloaders by their constructor function.

File: bootloader/bootloader.go (hunk 2 above)
Comment (bboozzoo) [reply bootloader/bootloader.go:327]: I'll add a TODO which we can address with the changes I have here https://github.com/bboozzoo/snapd/commit/9b684dfadba34afd7199891ba8bfe0d6ec56b1d6

File: bootloader/bootloader.go [hunk 6]
```diff
@@ -234,6 +235,19 @@ func InstallBootConfig(gadgetDir, rootDir string, opts *Options) error {
...
 }
 
+type bootloaderNewFunc func(rootdir string, opts *Options) Bootloader
+
+var (
+	//  bootloaders list all possible bootloaders by their constructor
+	//  function.
+	bootloaders = []bootloaderNewFunc{
+		newGrub,
+		newUboot,
```
Comment (anonymouse64) [inline bootloader/bootloader.go:245]: /me notes that this is note the same order as the original code, is this an issue? I don't think so, but just thought I would call it out

File: bootloader/bootloader.go [hunk 7]
```diff
@@ -327,3 +327,16 @@ func removeKernelAssetsFromBootDir(bootDir string, s snap.PlaceInfo) error {
 
 	return nil
 }
+
+// ForGadget returns a bootloader matching a given gadget by inspecting the
+// contents of gadget directory or en error if no matching bootloader is found.
```
Comment (anonymouse64) [inline bootloader/bootloader.go:332]:
```suggestion
// contents of gadget directory or an error if no matching bootloader is found.
```

File: bootloader/bootloader.go (hunk 6 above)
Comment (bboozzoo) [reply bootloader/bootloader.go]: I've pushed a patch that preserves the original order of uboot -> grub -> android -> lk.

**Pull request merged**